devices/src/virtio/net.rs

```diff
@@ -555,52 +559,54 @@ impl EpollHandler for NetEpollHandler {
             }
             TX_QUEUE_EVENT => {
                 METRICS.net.tx_queue_event_count.inc();
-                if let Err(e) = self.tx.queue_evt.read() {
-                    error!("Failed to get tx queue event: {:?}", e);
-                    METRICS.net.event_fails.inc();
-                    Err(DeviceError::FailedReadingQueue {
-                        event_type: "tx queue event",
-                        underlying: e,
+                self.tx
+                    .queue_evt
+                    .read()
+                    .map_err(|e| {
+                        error!("Failed to get tx queue event: {:?}", e);
+                        METRICS.net.event_fails.inc();
+                        DeviceError::FailedReadingQueue {
+                            event_type: "tx queue event",
+                            underlying: e,
+                        }
+                    })
+                    .and_then(|_| {
+                        if !self.tx.rate_limiter.is_blocked()
+                        // If the limiter is not blocked, continue transmitting bytes.
+                        {
+                            self.process_tx()
+                        } else {
+                            Ok(())
+                        }
                     })
-                } else if !self.tx.rate_limiter.is_blocked()
-                // If the limiter is not blocked, continue transmitting bytes.
-                {
-                    self.process_tx()
-                } else {
-                    Ok(())
-                }
             }
             RX_RATE_LIMITER_EVENT => {
                 METRICS.net.rx_event_rate_limiter_count.inc();
                 // Upon rate limiter event, call the rate limiter handler
                 // and restart processing the queue.
-                match self.rx.rate_limiter.event_handler() {
-                    Ok(_) => {
-                        // There might be enough budget now to receive the frame.
-                        self.resume_rx()
-                    }
-                    Err(e) => {
+                self.rx
```

Comment:
nit: same here, `match` is very rusty, changing `Ok()` and `Err()` branches of a `match` to `.map_err()` + `.and_then()` just makes the code harder to read with absolutely no added value, in my opinion.

Reply:
Performance-freak:

The `match` is also a single branch. `.map_err()` compiles down to a branch where the `Err()` case is diverted to the closure, otherwise code flows to `.and_then()` where you get a second branch where this time `Ok()` is diverted to the closure, otherwise code just flows.

I'm not sure the compiler knows to optimize this to a single branch/jump as it would in the `match` case.

Reply:
This should be a pretty straightforward optimization; if anyone wants to try something out, there are sites like https://rust.godbolt.org/ where you can see the assembly output for some Rust code (don't forget to supply the -O [as in capital "o"] option, otherwise you'll get a debug build).

Reply:
The generated code looks to be identical to me

1. [map_err + map](https://rust.godbolt.org/z/3lED-k)
3. [match](https://rust.godbolt.org/z/OjxsKT)

<img width="1910" alt="screenshot 2019-01-08 at 15 41 41" src="https://user-images.githubusercontent.com/914805/50814132-f8490f00-135b-11e9-9006-62452f8111bc.png">

Reply:
Since there are no performance advantages with either approach, let's go with the one that makes cleaner code :smiley:

Reply:
Sounds good to me. Reverted 97e9ffd0c862ceb0d79393b11b1a4a329e57eac3.